File: unknown.cpp
// TC:O(logn);
// SC:O(1);

// Approach:
// Here we find the ranges of search space by keeping intially low and high pointers at 0,1 indices
// checking whether high is less than target and updating high to high*2, And then we perform the 
// Binary search in this range;

class Solution {
public:
    int search(const ArrayReader& reader, int target) {

        int low=0;
        int high=1;
        // First step is to find the range of search space
        while(reader.get(high)<target)
        {
                low=high;
                high=high*2;            
        }
        // then performing Binary search on Search space
        return BS(reader,target,low,high);
    }

    int BS(const ArrayReader& reader, int target,int low,int high)
    {
        while(low<=high)
        {
            int mid=low+(high-low)/2;
            if(reader.get(mid)==target)
            return mid;
            else if(target<reader.get(mid))
            {
                high=mid-1;
            }
            else
            {
                low=mid+1;
            }
        }
        return -1;
    }
};

File: unknown.java
package s30;
interface ArrayReader{
	public static void get(int index) {}
}
public class unknown{

	public int search(ArrayReader reader, int target) {
		if(reader.get(0)==target) return 0;
		int low=0;
		int high=1;
		while(reader.get(high) < target){
			low =high;
			high = 2*high;
		}
		while(low <= high) {
		int mid = low + (high -
		low)/2;
		if(reader.get(mid) == target){
		return mid;
		} 
		else if(reader.get(mid) > target) {
			high = mid - 1;
		} 
		else {
			low = mid + 1;
		}
		}
		return -1;
	}
}
